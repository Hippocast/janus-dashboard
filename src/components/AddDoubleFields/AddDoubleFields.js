import React from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';

import block from '../../helpers/bem-cn';
import parse from '../../helpers/parse-value';

import Row from '../../modules/Layout/Row/Row';
import Label from '../../components/Label/Label';
import Input from '../../modules/inputs/Input';
import Control from '../../components/Control/Control';

import './AddDoubleFields.css';

const row = block('j-row');

const propTypes = {
    name: PropTypes.string.isRequired,
    title: PropTypes.string,
    config: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const AddDoubleFields = ({ config, name, title }) => {
    const passParse = config => config.type === 'number' ? parse : undefined;
    const renderMembers = ({ fields, config, hint, title }) => {
        return (
            <div>
                <div className={row()}>
                    <Label>{ title }</Label>
                    <Control
                        onClick={() => fields.push({})}
                        icon="add"
                    />
                </div>
                {
                    fields.map((member, index) => (
                        <Row className="double-fields" key={index} col>
                            <div className={row()}>
                                <div className={row('item', {pair: true})()}>
                                    <Field
                                        name={`${member}.${config[0].sufix}`}
                                        type={config[0].type}
                                        component={Input}
                                        placeholder={config[0].placeholder}
                                        parse={passParse(config[0])}
                                    />
                                </div>
                                <div className={row('item')()}>
                                    <Field
                                        name={`${member}.${config[1].sufix}`}
                                        type={config[1].type}
                                        component={Input}
                                        placeholder={config[1].placeholder}
                                        parse={passParse(config[1])}
                                    />
                                </div>
                                <div className={row('control')()}>
                                    <Control
                                        onClick={() => fields.remove(index)}
                                        icon="remove"
                                    />
                                </div>
                            </div>
                        </Row>
                    ))
                }
            </div>
        );
    };

    return (
        <div className="j-col__item">
            <FieldArray
                name={`${name}`}
                component={renderMembers}
                title={title}
                config={config}
            />
        </div>
    );
};

AddDoubleFields.propTypes = propTypes;

export default AddDoubleFields;
