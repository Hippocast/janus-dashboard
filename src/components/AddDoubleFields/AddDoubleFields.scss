@import 'index.scss';

$padding-size: ceil($hf-gutter / 2);

.double-fields {
    padding-top: $padding-size;
    padding-bottom: $padding-size;
}
