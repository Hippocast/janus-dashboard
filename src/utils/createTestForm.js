import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';

const Form = ({ children, initialValues }) => children;

const createTestForm = initialValues => connect(
    () => ({
        initialValues
    }),
    null,
)(
    reduxForm({
        form: 'mockForm',
    })(Form)
);

export default createTestForm;
