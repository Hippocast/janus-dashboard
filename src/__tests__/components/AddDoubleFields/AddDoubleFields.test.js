import React from 'react';
import renderer from 'react-test-renderer';
import { reduxForm } from 'redux-form';
import { createStore } from 'redux';
import { connect, Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import createTestForm from '../../../utils/createTestForm';
import AddDoubleFields from '../../../components/AddDoubleFields/AddDoubleFields';

const initialValues = {
    'mock-name': [
        {
            name: 'me',
        },
        {
            name: 'you',
        },
    ]
};

const store = createStore(() => ({
    form: {
        mockForm: {
            values: initialValues,
        }
    }
}));

const ConnectedForm = createTestForm(initialValues);

const renderFakeForm = store => el =>
    <Provider store={store}>
        <ConnectedForm>
            {el}
        </ConnectedForm>
    </Provider>;

const wrap = store => el => renderer
    .create(
        renderFakeForm(store)(el)
    );

describe('AddDoubleFields component', () => {
    const requiredProps = {
        name: 'mock-name',
        title: 'mock-title',
        config: [
            {
                name: 'me',
            },
            {
                name: 'you',
            },
        ]
    };

    it('renders correctly', () => {
        const tree = wrap(store)(
            <AddDoubleFields
                {...requiredProps}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
