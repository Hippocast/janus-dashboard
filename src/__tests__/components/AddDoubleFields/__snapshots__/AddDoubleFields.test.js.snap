// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDoubleFields component renders correctly 1`] = `
<div
  className="j-col__item"
>
  <div>
    <div
      className="j-row"
    >
      <label
        className="j-label"
        htmlFor={undefined}
      >
        mock-title
      </label>
      <button
        className="j-control"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label={undefined}
          className="j-icon j-icon--type-add j-control__icon"
        />
      </button>
    </div>
    <div
      className="j-col__item double-fields"
    >
      <div
        className="j-row"
      >
        <div
          className={[Function]}
        >
          <input
            className="j-input"
            disabled={undefined}
            name="mock-name[0].undefined"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder={undefined}
            required={undefined}
            type={undefined}
            value=""
          />
        </div>
        <div
          className={[Function]}
        >
          <input
            className="j-input"
            disabled={undefined}
            name="mock-name[0].undefined"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder={undefined}
            required={undefined}
            type={undefined}
            value=""
          />
        </div>
        <div
          className="j-row__control"
        >
          <button
            className="j-control"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label={undefined}
              className="j-icon j-icon--type-remove j-control__icon"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="j-col__item double-fields"
    >
      <div
        className="j-row"
      >
        <div
          className={[Function]}
        >
          <input
            className="j-input"
            disabled={undefined}
            name="mock-name[1].undefined"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder={undefined}
            required={undefined}
            type={undefined}
            value=""
          />
        </div>
        <div
          className={[Function]}
        >
          <input
            className="j-input"
            disabled={undefined}
            name="mock-name[1].undefined"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder={undefined}
            required={undefined}
            type={undefined}
            value=""
          />
        </div>
        <div
          className="j-row__control"
        >
          <button
            className="j-control"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-label={undefined}
              className="j-icon j-icon--type-remove j-control__icon"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
